### PR TITLE
Use libp2p gossipsub upstream validator framework

### LINKED ISSUE #4317: Gossipsub Usage Is Incorrect
@protolambda has brought to light some problems with our current gossip sub usage. Currently we have two main problems with how we use it. 

1) Excessive Broadcasting 
We currently manually re-broadcast each message we receive; however the gossipsub router does this internally too over here. https://github.com/libp2p/go-libp2p-pubsub/blob/7e5b02f152ae2dd8bceb897affe7ef57cb3c6b4e/pubsub.go#L761 
So messages are being propagated twice, leading to message amplification occurring. This has been the cause of high average message counts.

2) Validation is separate from the gossipsub callback. We do message validation in our pipeline, not involving the gossipsub service of this. This leads to both good/bad messages being relayed to other peers. So whatever validation we do only affects the messages that we manually broadcast. We will need to add in the gossipsub validator to validate messages from specific topics

Related to #4218

### PR DESCRIPTION
**Pull request by prestonvanloon**
Resolves #4317 

Thanks to @protolambda for pointing this out. 

This is now a refactoring to use the gossipsub validator framework.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
This solution is not going to work as is. Adding a pubsub validator to reject all re-propagation will block all messages from being received.

The problem with using the pubsub validator framework for validation is that when you need to inspect the message contents, you would need to deserialize the message more than once for any given valid message. With the assumption that the majority of the messages will be valid, this could be an expensive task. 

As is, we all repropagate messages once, thanks to content addressable message IDs. The problem is that we want to stop propagation of messages that are invalid. 

My thinking now is that we consider these two approaches:
- Add some functionality in gossipsub to disable automatic propagation of any topic message
- Use the existing validator framework and deserialize messages twice. Alternatively, come up with some caching mechanism to try to avoid deserializing the messages twice, but this idea smells.

I've reached out to @raulk on telegram to see if there is any expert feedback from their team.

**Comment by protolambda** [comment]
There's another option: introduce a preprocessing step, to then continue with the preprocessed message in validation and the subscription.
I see three options for continuation there:
- overload the message interface that gets passed around, and decode it into the right typed data from preprocessing for validation/processing.
- add a metadata getter to the message. To attach things to during processing, to use at a later stage.
- refactor more, make the pipeline more compositional, and somehow progressively process the message at each stage, instead of repeating things like decoding

**Comment by prestonvanloon** [comment]
>overload the message interface that gets passed around, and decode it into the right typed data from preprocessing for validation/processing.

> refactor more, make the pipeline more compositional, and somehow progressively process the message at each stage, instead of repeating things like decoding

These options would certainly solve our problem.

I imagine this pipeline:
1. Incoming pubsub.Message
1. Check seen message ID, peer signature verification, other pubsub metadata checks
1. Preprocess -> decode message
1. Validate -> check message contents
1. Propagate
1. Subscriber receives message -> handle message at application layer

**Comment by raulk** [comment]
Yes, I agree that we could add a baggage map where validators could store deserialised payloads or other computed/derived information, for downstream validator to reuse. We could also propagate this baggage to the final consumer.

**Comment by whyrusleeping** [comment]
Oh hey! we already have a place for the validators to put things for the consumer, you can set Message.ValidatorData in the validator, and it is accessible by the receiver.

ref: https://github.com/libp2p/go-libp2p-pubsub/pull/231

**Comment by raulk** [comment]
Yep, it turns out this is already implemented and I had missed it :-)

**Comment by prestonvanloon** [comment]
Awesome. Thanks @whyrusleeping @raulk @protolambda. We'll refactor our validation pipeline and use that field to pass around the message data. Thanks!!!

**Comment by prestonvanloon** [comment]
OK this is ready for review again. I was experiencing some weird behavior that appears to run a validator on broadcast as well as the subscriber. https://github.com/libp2p/go-libp2p-pubsub/issues/250

If anyone knows how to disable the validator on publish, I'd appreciate that info.

File: beacon-chain/sync/validate_beacon_attestation.go [hunk 1]
```diff
@@ -32,55 +34,55 @@ func (r *Service) validateBeaconAttestation(ctx context.Context, msg proto.Messa
 	// TODO(1332): Add blocks.VerifyAttestation before processing further.
 	// Discussion: https://github.com/ethereum/eth2.0-specs/issues/1332
 
-	att, ok := msg.(*ethpb.Attestation)
+	if msg == nil || msg.TopicIDs == nil || len(msg.TopicIDs) == 0 {
+		return false
+	}
+	topic := msg.TopicIDs[0]
+	topic = strings.TrimSuffix(topic, r.p2p.Encoding().ProtocolSuffix())
```
[the remark this answers is not in the document]
Comment (protolambda) [reply beacon-chain/sync/validate_beacon_attestation.go]: Nice refactor, ninja'd this problem before I submitted the review. Great

File: beacon-chain/sync/subscriber.go [hunk 2]
```diff
@@ -4,37 +4,40 @@ import (
...
-// pipeline. FromSelf indicates whether or not this is a message received from our
-// node in pubsub.
-type validator func(ctx context.Context, msg proto.Message, broadcaster p2p.Broadcaster, fromSelf bool) (bool, error)
-
-// noopValidator is a no-op that always returns true and does not propagate any
-// message.
-func noopValidator(_ context.Context, _ proto.Message, _ p2p.Broadcaster, _ bool) (bool, error) {
-	return true, nil
+// noopValidator is a no-op that only decodes the message, but does not check its contents.
+func (r *Service) noopValidator(ctx context.Context, _ peer.ID, msg *pubsub.Message) bool {
```
[the remark this answers is not in the document]
Comment (protolambda) [reply beacon-chain/sync/subscriber.go]: Also, I think pubsub has a way to chain validators in sync, so you do not have to repeatedly call decodePubsubMessage, and put this as the preprocess validator instead.

File: beacon-chain/sync/validate_beacon_attestation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/sync/validate_beacon_attestation.go]: Thanks! Terence beat you to it lol

File: beacon-chain/sync/subscriber.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/sync/subscriber.go]: This is actually only used in test. From what I recall, all validators were run asynchronously. Chaining would be nice with your suggestion though. 

From godocs: 

> RegisterTopicValidator registers a validator for topic. By default validators are asynchronous, which means they will run in a separate goroutine. The number of active goroutines is controlled by global and per topic validator throttles; if it exceeds the throttle threshold, messages will be dropped.

I'll leave this for now unless we can chain these.

File: beacon-chain/sync/validate_attester_slashing.go [hunk 3]
```diff
@@ -3,96 +3,60 @@ package sync
...
 		}
 	}
 
 	if err := blocks.VerifyAttesterSlashing(ctx, s, slashing); err != nil {
-		seenAttesterSlashings.Set(invalidKey, true /*value*/, 1 /*cost*/)
-		return false, errors.Wrap(err, "Received invalid attester slashing")
-	}
-	seenAttesterSlashings.Set(cacheKey, true /*value*/, 1 /*cost*/)
-
-	if fromSelf {
```
Comment (nisdas) [inline beacon-chain/sync/validate_attester_slashing.go]: does libp2p not process messages from self ?

Comment (prestonvanloon) [reply beacon-chain/sync/validate_attester_slashing.go]: That's right, it won't send messages from itself to the subscription. It does, however, run validation on all messages published or subscribed. I've added a bit of a hack to immediately return true in validation if the pubsub message comes from ourselves. This bypasses the validation effectively as suggested here https://github.com/libp2p/go-libp2p-pubsub/issues/250#issuecomment-567746609

**Pull request merged**